### PR TITLE
Add RequestRejectedHandler

### PR DESCRIPTION
**Pull request by leonard84**
…uests

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@leonard84 Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=7052)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-security&pullRequestId=7052) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by leonard84** [comment]
@rwinch the company cla is currently at our legal department, but you can review the pr already.

File: web/src/main/java/org/springframework/security/web/FilterChainProxy.java [hunk 1]
```diff
@@ -190,29 +198,33 @@ public void doFilter(ServletRequest request, ServletResponse response,
 	private void doFilterInternal(ServletRequest request, ServletResponse response,
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/FilterChainProxy.java]: Rather than having this entire method be between a try/catch, can you please move the try/catch to the existing try block in `doFilter`?

File: web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * A simple implementation of {@link RequestRejectedHandler} that sends and error with status code 400.
+ */
+public class ClientErrorRequestRejectedHandler implements RequestRejectedHandler {
+	@Override
+	public void handle(HttpServletRequest request, HttpServletResponse response,
+			RequestRejectedException requestRejectedException) throws IOException {
+		response.sendError(HttpServletResponse.SC_BAD_REQUEST);
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java]: I think we should add the exception message to sendError so that the user knows what went wrong.

File: web/src/main/java/org/springframework/security/web/FilterChainProxy.java [hunk 3]
```diff
@@ -272,6 +284,15 @@ public void setFirewall(HttpFirewall firewall) {
 		this.firewall = firewall;
 	}
 
+	/**
+	 * Sets the {@link RequestRejectedHandler} to be used for requests rejected by the firewall.
+	 * 
+	 * @param requestRejectedHandler the {@link RequestRejectedHandler}
+	 */
+	public void setRequestRejectedHandler(RequestRejectedHandler requestRejectedHandler) {
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/FilterChainProxy.java]: Please make sure this is not null and add a test for it

File: web/src/main/java/org/springframework/security/web/firewall/RequestRejectedHandler.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Used by {@link org.springframework.security.web.FilterChainProxy} to handle an
+ * <code>RequestRejectedException</code>.
+ *
+ * @author Leonard Brünings
+ */
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/firewall/RequestRejectedHandler.java]: Please add `@since 5.2`

File: web/src/main/java/org/springframework/security/web/firewall/DefaultRequestRejectedHandler.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package org.springframework.security.web.firewall;
+
+import java.io.IOException;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Default implementation of {@link RequestRejectedHandler} that simply rethrows the exception.
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/firewall/DefaultRequestRejectedHandler.java]: Please add `@since 5.2`

File: web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package org.springframework.security.web.firewall;
+
+import java.io.IOException;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * A simple implementation of {@link RequestRejectedHandler} that sends and error with status code 400.
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java]: Please add `@since 5.2`

File: web/src/main/java/org/springframework/security/web/FilterChainProxy.java [hunk 7]
```diff
@@ -272,6 +284,15 @@ public void setFirewall(HttpFirewall firewall) {
 		this.firewall = firewall;
 	}
 
+	/**
+	 * Sets the {@link RequestRejectedHandler} to be used for requests rejected by the firewall.
+	 * 
+	 * @param requestRejectedHandler the {@link RequestRejectedHandler}
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/FilterChainProxy.java]: Please add `@since 5.2`

File: web/src/main/java/org/springframework/security/web/FilterChainProxy.java [hunk 8]
```diff
@@ -16,27 +16,33 @@
 
 package org.springframework.security.web;
 
+import java.io.IOException;
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/FilterChainProxy.java]: Please remove any unnecessary changes (i.e. formatting) that are not related to the issue.

File: web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java (hunk 2 above)
Comment (leonard84) [reply web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java]: Are we sure that the message does not leak any security relevant information?

File: web/src/main/java/org/springframework/security/web/FilterChainProxy.java (hunk 8 above)
Comment (leonard84) [reply web/src/main/java/org/springframework/security/web/FilterChainProxy.java]: Yeah IDE auto-format, is there a profile I can import? Otherwise it is a PITA to try to get IntelliJ to leave them alone.

**Comment by leonard84** [comment]
@rwinch I had some open questions in the issue:

* `FilterChainProxy` is configured in `WebSecurity`, but the `ExceptionHandlingConfigurer` lives in a different package, so the `getRequestRejectedHandler()` would have to be public. Would it make more sense to make it a property of `WebSecurity`?
* Travis seems to have removed Oracle Java 8 support

```
$ ~/bin/install-jdk.sh --target "/home/travis/oraclejdk8" --workspace "/home/travis/.cache/install-jdk" --feature "8" --license "BCL"
install-jdk.sh 2019-06-25
Expected feature release number in range of 9 to 14, but got: 8
The command "~/bin/install-jdk.sh --target "/home/travis/oraclejdk8" --workspace "/home/travis/.cache/install-jdk" --feature "8" --license "BCL"" failed and exited with 3 during .
```

File: web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java (hunk 2 above)
Comment (leonard84) [reply web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java]: I'm adding logging here, should it be `WARN`, `INFO`, or `DEBUG`?
```java
if (logger.isDebugEnabled()) {
	logger.debug("Rejecting request due to: "+requestRejectedException.getMessage(),
			requestRejectedException);
}
```

Comment (rwinch) [reply web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java]: Debug please.

**Comment by leonard84** [comment]
@rwinch friendly ping, see above

**Comment by rwinch** [comment]
Thanks for the ping.

I'd say we should just expose the handler via a Bean lookup similar to how httpFirewall is looked up in `WebSecurity`.

I'm not sure what is happening with the JDK. Our travis build is passing with JDK 8. How did you reproduce this?

**Comment by rwinch** [comment]
FYI it appears the Travis build was related to gh-7169 which has been resolved

**Comment by pivotal-issuemaster** [comment]
@leonard84 Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=7052)!

**Comment by leonard84** [comment]
@rwinch please take another look

File: web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+/**
+ * A simple implementation of {@link RequestRejectedHandler} that sends and error with status code 400.
+ *
+ * @author Leonard Brünings
+ * @since 5.2
+ */
+public class ClientErrorRequestRejectedHandler implements RequestRejectedHandler {
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java]: Please add tests for this class

File: web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+	private static final Log logger = LogFactory.getLog(ClientErrorRequestRejectedHandler.class);
+
+	@Override
+	public void handle(HttpServletRequest request, HttpServletResponse response,
+			RequestRejectedException requestRejectedException) throws IOException {
+		if (logger.isDebugEnabled()) {
+			logger.debug("Rejecting request due to: " + requestRejectedException.getMessage(),
+					requestRejectedException);
+		}
+		response.sendError(HttpServletResponse.SC_BAD_REQUEST);
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/firewall/ClientErrorRequestRejectedHandler.java]: I think we should allow the http status to be injected through the constructor rather than hard coding the status. We should also rename the class to reflect this change. Finally, a default constructor with bad request as the default also makes a lot of sense

File: web/src/main/java/org/springframework/security/web/firewall/DefaultRequestRejectedHandler.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * @author Leonard Brünings
+ * @since 5.2
+ */
+public class DefaultRequestRejectedHandler implements RequestRejectedHandler {
```
Comment (rwinch) [inline web/src/main/java/org/springframework/security/web/firewall/DefaultRequestRejectedHandler.java]: Please add a test

**Comment by leonard84** [comment]
@rwinch I've implemented your requests

**Comment by leonard84** [comment]
@rwinch ping

**Comment by leonard84** [comment]
@rwinch ping again

**Comment by leonard84** [comment]
@rwinch I've rebased to fix the merge problems and squashed the commits into one

**Comment by skjolber** [comment]
@rwinch ping

**Comment by ptahchiev** [comment]
Ping again - this is a much needed feature :)

**Comment by patrykks** [comment]
Any update in this thread? This is really needed by community.

File: web/src/main/java/org/springframework/security/web/FilterChainProxy.java [hunk 12]
```diff
@@ -176,6 +182,8 @@ public void doFilter(ServletRequest request, ServletResponse response,
 			try {
 				request.setAttribute(FILTER_APPLIED, Boolean.TRUE);
 				doFilterInternal(request, response, chain);
+			} catch (RequestRejectedException e) {
```
Comment (RomanGrytsay) [inline web/src/main/java/org/springframework/security/web/FilterChainProxy.java:185]: Does this need to be handled in the `else` block (line 194) as well?

File: web/spring-security-web.gradle [hunk 13]
```diff
@@ -26,6 +26,7 @@ dependencies {
 	testCompile 'org.skyscreamer:jsonassert'
 	testCompile 'org.springframework:spring-webflux'
 	testCompile 'org.synchronoss.cloud:nio-multipart-parser'
+	testCompile 'cglib:cglib-nodep'
```
Comment (rwinch) [inline web/spring-security-web.gradle:29]: Is there a reason this is necessary?

File: web/src/main/java/org/springframework/security/web/FilterChainProxy.java (hunk 12 above)
Comment (leonard84) [reply web/src/main/java/org/springframework/security/web/FilterChainProxy.java:185]: I don't think so, as it has already been applied in the filter chain before as indicated by the `FILTER_APPLIED` attribute being present, so it is already in a surrounding try-catch, doing it again could call the handler more than once.

File: web/spring-security-web.gradle (hunk 13 above)
Comment (leonard84) [reply web/spring-security-web.gradle]: Not anymore since byte-buddy is used

**Comment by leonard84** [comment]
@rwinch I've rebased and converted the Spock tests to the old JUnit 4 style and removed the cglib dependency as byte-buddy is now used to proxy classes. 
Out of interest, what was the reason for the removal of Spock/groovy for testing?

**Comment by rwinch** [comment]
Thanks for the PR @leonard84! This is now merged into master. I've added a bit of polish 0483b3e0421b1fdf55ac5d147d999c6c89986206 and support for configuring via XML 4a9fa0337a5eb5dc2603afa45fd513fdf001aaea.

> Out of interest, what was the reason for the removal of Spock/groovy for testing?

As a Java project, testing in Groovy made an unnecessary hurtle for contributors. We switched to make contributions easier for our audience.
